02-peer-protocol.md

```diff
@@ -50,10 +67,361 @@ pubkey corresponding to the funding output nothing prevents duplicative channel
+interactive transaction construction protocol is considered concluded.
+Both peers should construct the transaction and fail the negotiation
+if an error is discovered.
+
+This protocol is expressly designed to allow for parallel, multi-party
+sessions to collectively construct a single transaction. This preserves
+the ability to open multiple channels in a single transaction. While
+`serial_id`s are generally chosen randomly, to maintain consistent transaction
+ordering across all peer sessions, it is simplest to invert the bottom-bit of
+received `serial_id` before forwarding them to other peers.
+
+Here are a few example exchanges.
+
+#### *initiator* only
+
+A, *initiator* has two inputs and an output (the funding output).
+B, the *non-initiator* has nothing to contribute.
+n
+
+        +-------+                       +-------+
+        |       |--(1)- tx_add_input -->|       |
+        |       |<-(2)- tx_complete ----|       |
+        |       |--(3)- tx_add_input -->|       |
+        |   A   |<-(4)- tx_complete ----|   B   |
+        |       |--(5)- tx_add_output ->|       |
+        |       |                       |       |
+        |       |--(6)- tx_complete --->|       |
+        |       |<-(7)- tx_complete ----|       |
+        +-------+                       +-------+
+
+#### *initiator* and *non-initiator*
+
+A the *initiator* contributes 2 inputs and an output that they
+then remove.  B, the *non-initiator*, contributes 1 input and an output,
+but waits until A adds a second input before contributing.
+
+Note that if A does not send a second input, the negotiation will end without
+B's contributions.
+
+        +-------+                       +-------+
+        |       |--(1)- tx_add_input -->|       |
+        |       |<-(2)- tx_complete ----|       |
+        |       |--(3)- tx_add_output ->|       |
+        |       |<-(4)- tx_complete ----|       |
+        |       |--(5)- tx_add_input -->|       |
+        |   A   |<-(6)- tx_add_input ---|   B   |
+        |       |<-(7)- tx_add_output --|       |
+        |       |--(8)- tx_rm_output -->|       |
+        |       |<-(9)- tx_complete ----|       |
+        |       |--(10) tx_complete --->|       |
+        +-------+                       +-------+
+
+
+### The `tx_add_input` Message
+
+This message contains a transaction input.
+
+1. type: 66 (`tx_add_input`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u64`:`serial_id`]
+    * [`u16`:`prevtx_len`]
+    * [`prevtx_len*byte`:`prevtx`]
+    * [`u32`:`prevtx_vout`]
+    * [`u32`:`sequence`]
+    * [`u16`:`script_sig_len`]
+    * [`script_sig_len*byte`:`script_sig`]
+
+#### Requirements
+
+The sending node:
+  - MUST add all sent inputs to the transaction
+  - MUST use a unique `serial_id` for each input currently added to the
+    transaction
+  - MUST NOT re-transmit inputs it has received from the peer
+  - if is the *initiator*:
+    - MUST send even `serial_id`s
+  - if is the *non-initiator*:
```

Comment:
I think I would uplift *initiator*/*non-initiator* in its own section. Those requirements play out beyond `tx_add_input` ? Also *initiator* MUST contribute at least one input ?

Reply:
There's no concrete reason the initiator must contribute one input; in the case of the funding transaction this is true and has been added to that particular section of the document. For a closing transaction, however, *one* of the peers must add the funding output as part of the negotiation but there's no reason it necessarily MUST be the initiator (it is sane to require that they do so for that particular protocol however).

Although repeated between `tx_add_input` and `tx_add_output`, it was thought more practical/useful to the implementer to have the complete set of requirements listed explicitly for each message.